File: app/src/main/java/com/like/view/LikeStar.java
package com.like.view;

import android.animation.TypeEvaluator;
import android.animation.ValueAnimator;
import android.content.Context;
import android.graphics.PointF;
import android.graphics.drawable.Drawable;
import android.util.AttributeSet;
import android.view.View;
import android.view.ViewGroup;
import android.view.animation.AccelerateDecelerateInterpolator;
import android.view.animation.AccelerateInterpolator;
import android.view.animation.DecelerateInterpolator;
import android.view.animation.Interpolator;
import android.view.animation.LinearInterpolator;
import android.widget.ImageView;

import com.like.R;

import java.util.ArrayList;
import java.util.List;
import java.util.Random;

public class LikeStar extends ViewGroup {
    private List<Drawable> mStarDrawable;
    private int mWidth;
    private int mHeight;
    //定义贝塞尔曲线的数据点和两个控制点
    private PointF startP, endP, pointX, pointY;
    private Random random = new Random();
    public LikeStar(Context context) {
        this(context, null);
    }

    public LikeStar(Context context, AttributeSet attrs) {
        this(context, attrs, 0);
    }

    public LikeStar(Context context, AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
        init(context);
    }

    private void init(final Context context) {
        mStarDrawable = new ArrayList<>();
        startP = new PointF();
        endP = new PointF();
        pointX = new PointF();
        pointY = new PointF();
        //初始化图片资源
        mStarDrawable.add(getResources().getDrawable(R.drawable.like_red, null));
        mStarDrawable.add(getResources().getDrawable(R.drawable.like_blue, null));
        mStarDrawable.add(getResources().getDrawable(R.drawable.like_green, null));
        mStarDrawable.add(getResources().getDrawable(R.drawable.like_repple, null));
        //初始化插补器
        @SuppressWarnings("MismatchedQueryAndUpdateOfCollection")
        List<Interpolator> interpolators = new ArrayList<>();
        interpolators.add(new LinearInterpolator());
        interpolators.add(new AccelerateDecelerateInterpolator());
        interpolators.add(new AccelerateInterpolator());
        interpolators.add(new DecelerateInterpolator());

        ImageView image_heard = new ImageView(context);
        image_heard.setImageDrawable(mStarDrawable.get(0));

        image_heard.setLayoutParams(new LayoutParams(LayoutParams.WRAP_CONTENT,
                LayoutParams.WRAP_CONTENT));
        image_heard.setOnClickListener(v -> {
            //点击之后开始动画,添加红心到布局文件并开始动画
            final ImageView image_random = new ImageView(context);
            image_random.setImageDrawable(mStarDrawable.get(random.nextInt(4)));

            image_random.setLayoutParams(new LayoutParams(LayoutParams.WRAP_CONTENT,
                    LayoutParams.WRAP_CONTENT));
            addView(image_random);
            invalidate();
            //开始做动画效果
            PointF endPointRandom = new PointF(random.nextInt(mWidth), endP.y);
            BezierTypeEvaluator bezierTypeEvaluator = new BezierTypeEvaluator(new PointF(random.nextInt(mWidth), random.nextInt(mHeight)), new PointF(random.nextInt(mWidth), random.nextInt(mHeight)));
            ValueAnimator valueAnimator = ValueAnimator.ofObject(bezierTypeEvaluator, startP, endPointRandom);
            valueAnimator.addUpdateListener(animation -> {
                PointF pointF = (PointF) animation.getAnimatedValue();
                image_random.setX(pointF.x);
                image_random.setY(pointF.y);
            });

            valueAnimator.setDuration(2000);
            valueAnimator.start();

        });
        addView(image_heard);
    }

    @Override
    protected void onSizeChanged(int w, int h, int oldW, int oldH) {
        super.onSizeChanged(w, h, oldW, oldH);
        mWidth = getMeasuredWidth();
        mHeight = getMeasuredHeight();
        // 初始化各个点
        //借用第一个子view控件中的宽高
        View child = getChildAt(0);
        int childW = child.getMeasuredWidth();
        int childH = child.getMeasuredHeight();

        startP.x = (mWidth - childW) >> 1;
        startP.y = mHeight - childH;
        endP.x = (mWidth - childW) >> 1;
        endP.y = 0 - childH;

        pointX.x = random.nextInt(mWidth / 2);
        pointX.y = random.nextInt(mHeight / 2) + (mHeight >> 1);

        pointY.x = random.nextInt(mWidth / 2) + (mWidth >> 1);
        pointY.y = random.nextInt(mHeight / 2);
    }

    @Override
    protected void onMeasure(int widthMeasureSpec, int heightMeasureSpec) {
        measureChildren(widthMeasureSpec, heightMeasureSpec);
        //获取view的宽高测量模式
        int widthSize = MeasureSpec.getSize(widthMeasureSpec);
        int heightSize = MeasureSpec.getSize(heightMeasureSpec);
        //保存测量高度
        setMeasuredDimension(widthSize, heightSize);
    }

    @Override
    protected void onLayout(boolean changed, int l, int t, int r, int b) {
        for (int i = 0; i < getChildCount(); i++) {
            View child = getChildAt(i);
            int childW = child.getMeasuredWidth();
            int childH = child.getMeasuredHeight();
            child.layout((mWidth - childW) / 2, (mHeight - childH), (mWidth - childW) / 2 + childW, mHeight);
        }
    }

    public class BezierTypeEvaluator implements TypeEvaluator<PointF> {
        private PointF pf1, pf2;

        BezierTypeEvaluator(PointF pf1, PointF pf2) {
            this.pf1 = pf1;
            this.pf2 = pf2;
        }

        @Override
        public PointF evaluate(float fraction, PointF startValue, PointF endValue) {
            PointF pointCur = new PointF();
            pointCur.x = startP.x * (1 - fraction) * (1 - fraction) * (1 - fraction) + 3
                    * pf1.x * fraction * (1 - fraction) * (1 - fraction) + 3
                    * pf2.x * (1 - fraction) * fraction * fraction + endValue.x * fraction * fraction * fraction;// 实时计算最新的点X坐标
            pointCur.y = startP.y * (1 - fraction) * (1 - fraction) * (1 - fraction) + 3
                    * pf1.y * fraction * (1 - fraction) * (1 - fraction) + 3
                    * pf2.y * (1 - fraction) * fraction * fraction + endValue.y * fraction * fraction * fraction;// 实时计算最新的点Y坐标
            return pointCur;
        }
    }
}

File: app/src/main/java/com/like/LikeStar.kt
package com.like

import android.animation.TypeEvaluator
import android.animation.ValueAnimator.ofObject
import android.content.Context
import android.graphics.PointF
import android.graphics.drawable.Drawable
import android.util.AttributeSet
import android.view.View.MeasureSpec.getSize
import android.view.ViewGroup
import android.view.ViewGroup.LayoutParams.WRAP_CONTENT
import android.view.animation.*
import android.widget.ImageView
import java.lang.Math.pow
import java.util.*

class LikeStar @JvmOverloads constructor(context: Context, attrs: AttributeSet, attr: Int = 0) :
        ViewGroup(context, attrs, attr) {
    private lateinit var starDrawable: MutableList<Drawable>
    private var widthM: Int = 0
    private var heightM: Int = 0
    //定义贝塞尔曲线的数据点和两个控制点
    private lateinit var startP: PointF
    private lateinit var endP: PointF
    private lateinit var pointX: PointF
    private lateinit var pointY: PointF
    private val random = Random()

    init {
        init(context)
    }

    private fun init(context: Context) {
        starDrawable = ArrayList() //初始化图片资源
        startP = PointF()
        endP = PointF()
        pointX = PointF()
        pointY = PointF()
        starDrawable.add(resources.getDrawable(R.drawable.like_red, null))
        starDrawable.add(resources.getDrawable(R.drawable.like_blue, null))
        starDrawable.add(resources.getDrawable(R.drawable.like_green, null))
        starDrawable.add(resources.getDrawable(R.drawable.like_repple, null))
        starDrawable.add(resources.getDrawable(R.drawable.like_orange, null))
        starDrawable.add(resources.getDrawable(R.drawable.like_yellow, null))
        val interpolators = ArrayList<Interpolator>() //初始化插补器
        interpolators.add(LinearInterpolator())
        interpolators.add(AccelerateDecelerateInterpolator())
        interpolators.add(AccelerateInterpolator())
        interpolators.add(DecelerateInterpolator())
        val defaultIcon = ImageView(context)
        defaultIcon.setImageDrawable(starDrawable[0])
        defaultIcon.layoutParams = ViewGroup.LayoutParams(WRAP_CONTENT, WRAP_CONTENT)
        defaultIcon.setOnClickListener {
            val iconRandom = ImageView(context) //点击之后开始动画,添加红心到布局文件并开始动画
            iconRandom.setImageDrawable(starDrawable[random.nextInt(6)])
            iconRandom.layoutParams = ViewGroup.LayoutParams(WRAP_CONTENT, WRAP_CONTENT)
            addView(iconRandom)
            invalidate()
            val bezier = Bezier(
                    PointF(random.nextInt(widthM).toFloat(), random.nextInt(heightM).toFloat()),
                    PointF(random.nextInt(widthM).toFloat(), random.nextInt(heightM).toFloat())
            ) //开启动画效果
            val anim = ofObject(bezier, startP, PointF(random.nextInt(widthM).toFloat(), endP.y))
            anim.addUpdateListener { animation ->
                val pointF = animation.animatedValue as PointF
                iconRandom.x = pointF.x
                iconRandom.y = pointF.y
            }
            anim.duration = 2000
            anim.start()
        }
        addView(defaultIcon)
    }

    override fun onSizeChanged(w: Int, h: Int, oldW: Int, oldH: Int) {
        super.onSizeChanged(w, h, oldW, oldH)  //初始化各点,借用第一个子控件中的宽高
        widthM = measuredWidth
        heightM = measuredHeight
        startP.x = (widthM - getChildAt(0).measuredWidth shr 1).toFloat()
        startP.y = (heightM - getChildAt(0).measuredHeight).toFloat()
        endP.x = (widthM - getChildAt(0).measuredWidth shr 1).toFloat()
        endP.y = (0 - getChildAt(0).measuredHeight).toFloat()
        pointX.x = random.nextInt(widthM / 2).toFloat()
        pointX.y = (random.nextInt(heightM / 2) + (heightM shr 1)).toFloat()
        pointY.x = (random.nextInt(widthM / 2) + (widthM shr 1)).toFloat()
        pointY.y = random.nextInt(heightM / 2).toFloat()
    }

    override fun onMeasure(widthMeasureSpec: Int, heightMeasureSpec: Int) {
        measureChildren(widthMeasureSpec, heightMeasureSpec) //获取控件宽高测量模式
        setMeasuredDimension(getSize(widthMeasureSpec), getSize(heightMeasureSpec)) //保存测量高度
    }

    override fun onLayout(changed: Boolean, l: Int, t: Int, r: Int, b: Int) {
        for (i in 0 until childCount) getChildAt(i).layout((widthM - getChildAt(i)
                .measuredWidth) / 2, heightM - getChildAt(i).measuredHeight, (widthM - getChildAt(i)
                .measuredWidth) / 2 + getChildAt(i).measuredWidth, heightM)
    }

    inner class Bezier(private val pf1: PointF, private val pf2: PointF) : TypeEvaluator<PointF> {
        override fun evaluate(f: Float, start: PointF, endValue: PointF): PointF {
            val pointCur = PointF()
            pointCur.x = startP.x * pow((1.0 - f), 3.0).toFloat() +
                    (3F * pf1.x * f * (1 - f) * (1 - f)) +
                    (3F * pf2.x * (1 - f) * f * f) +
                    endValue.x * f * f * f //实时计算最新的点X坐标
            pointCur.y = startP.y * pow((1.0 - f), 3.0).toFloat() +
                    (3F * pf1.y * f * (1 - f) * (1 - f)) +
                    (3F * pf2.y * (1 - f) * f * f) +
                    endValue.y * f * f * f //实时计算最新的点Y坐标
            return pointCur
        }
    }
}
